src/App/Fossa/VSI/DynLinked/Internal/Binary.hs

```diff
@@ -0,0 +1,132 @@
+module App.Fossa.VSI.DynLinked.Internal.Binary (
+  dynamicLinkedDependencies,
+  lddParseLocalDependencies,
+  lddParseDependency,
+  LocalDependency (..),
+) where
+
+import Control.Algebra (Has)
+import Control.Effect.Diagnostics (Diagnostics)
+import Control.Monad (void)
+import Data.Char (isSpace)
+import Data.Maybe (catMaybes)
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.String.Conversion (toText)
+import Data.Text (Text)
+import Data.Void (Void)
+import Effect.Exec (AllowErr (Never), Command (..), Exec, execParser)
+import Path (Abs, File, Path, parent, parseAbsFile, toFilePath)
+import System.Info qualified as SysInfo
+import Text.Megaparsec (Parsec, between, empty, eof, many, satisfy, try, (<|>))
+import Text.Megaparsec.Char (char, space1)
+import Text.Megaparsec.Char.Lexer qualified as L
+
+-- | Report the list of dynamically linked dependencies of the target executable as a set of file paths.
+-- This is currently a stub for non-linux targets: on these systems an empty set is reported.
+dynamicLinkedDependencies :: (Has Diagnostics sig m, Has Exec sig m) => Path Abs File -> m (Set (Path Abs File))
+dynamicLinkedDependencies file | SysInfo.os == "linux" = do
+  deps <- execParser lddParseLocalDependencies (parent file) $ Command "ldd" [toText $ toFilePath file] Never
+  pure . Set.fromList $ map localDependencyPath deps
+dynamicLinkedDependencies _ = pure Set.empty
+
+type Parser = Parsec Void Text
+
+data LocalDependency = LocalDependency
+  { localDependencyName :: Text
+  , localDependencyPath :: (Path Abs File)
+  }
+  deriving (Show, Eq, Ord)
+
+-- | Parse output from 'ldd' into a list of dependencies on disk.
+--
+-- Output from 'ldd' looks like this:
+--
+-- >  linux-vdso.so.1 =>  (0x00007ffc28d59000)
+-- >  libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fbea9a88000)
+-- >  /lib64/ld-linux-x86-64.so.2 (0x00007fbea9e52000)
+--
+-- Or this:
+--
+-- >  /lib/ld-musl-x86_64.so.1 (0x7f2ca52da000)
+-- >  libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x7f2ca52da000)
+--
+-- Each line is in the form @{name}{literal =>}{path}{memory address}@.
+--   * The memory address is ignored.
+--   * We consider each line to form a tuple of @({name}, {path})@.
+--     * Any line that doesn't form a complete tuple is ignored.
+--
+-- The above rules mean that the output above would evaluate to the following result:
+--
+-- > Set.fromList [$(mkAbsFile "/lib/x86_64-linux-gnu/libc.so.6")]
+lddParseLocalDependencies :: Parser [LocalDependency]
+lddParseLocalDependencies =
+  catMaybes
+    <$> many
```

Comment:
[Optional] You can also use `optional`, this way you can remove `lddConsumeSyscallLib`, `lddConsumeLinker`. This may make the code easier to read.

```haskell
optional :: Alternative f => f a -> f (Maybe a)
optional p = (Just <$> p) <|> pure Nothing
```

```haskell
import Text.Megaparsec (optional)

-- I have not tested this, but it should work.
lddParseLocalDependencies :: Parser [LocalDependency]
lddParseLocalDependencies = catMaybes <$> many (optional $ try lddParseDependency) <* eof

lddParseDependency :: Parser (LocalDependency)
lddParseDependency = (LocalDependency <$> (linePrefix *> ident) <* symbol "=>" <*> path <* printedHex)
```

Reply:
I'd love to do that, but we need to actually consume the lines we want to ignore: `optional $ try` endlessly retries to parse the same line 😞